File: src/ar/edu/ub/testing/rps/Round.java
package ar.edu.ub.testing.rps;

import java.util.Collection;
import java.util.ArrayList;
import java.util.HashMap;


/** A class that wraps a round of Rock, Paper, Scissors (or whatever the elements in play are) between two players. */
public class Round
{
    /** Initialize the round with the given players and playable elements.
     *
     * Args:
     *    playerA: A player in the form of a NamedObject (players do not have any other property).
     *    playerB: Another player that will play against the first one. Cannot be the same as playerA.
     *    availableElements: A list of Elements that are available for the players to pick while playing.
     */
    public Round(NamedObject playerA, NamedObject playerB, ArrayList<Element> availableElements)
    {
        this.plays = new HashMap<String, Element>();
        this.elements = new HashMap<String, Element>();

        if (playerA != null && playerB != null && playerB != playerA && !playerA.name().equals(playerB.name()))
        {
            this.playerA = playerA;
            this.playerB = playerB;
        }
        
        if (availableElements != null)
        {
            for (Element element : availableElements)
            {
                if (element != null && !this.elements.containsKey(element.name()) && !this.elements.values().contains(elements))
                {
                    this.elements.put(element.name(), element);
                }
            }
        }
    }

    /** Getter for the players that play against each other on the current Round.
     *  
     * Args:
     *   byName: Name of the player to get its associated object.
     *
     * Returns:
     *    NamedObject: A reference to the player being queried or null if not found.
     */
    public NamedObject player(String byName)
    {
        if (this.playerA.name().equals(byName))
        {
            return this.playerA;
        }
        else if (this.playerB.name().equals(byName))
        {
            return this.playerB;
        }
        return null;
    }

    /** The entire list of players that are part of the current Round.
     *   
     * Returns:
     *    ArrayList<NamedObject>: A list of references to players playing on this Round. If a player is missing,
     *        null will be present in the list.
     */
    public ArrayList<NamedObject> players()
    {
        ArrayList<NamedObject> players = new ArrayList<NamedObject>();
        players.add(this.playerA);
        players.add(this.playerB);
        return players;
    }

    /** Retrieve all Elements that are available for the players to play during this Round.
     *   
     * Returns:
     *    Collection<Element>: The list of Elements the players can choose from to play on this Round.
     */
    public Collection<Element> elements()
    {
        return this.elements.values();
    }

    /** Analog to player(String byName), allows retrieving a specific Element based on its name.
     *   
     * Args:
     *    byName: The name of the Element that should be looked up.
     *
     * Returns:
     *    Element: The Element reference, if found, or null if not.
     */
    public Element element(String byName)
    {
        return this.elements.get(byName);
    }

    /** Allows a player to perform a play.
     *   
     * Args:
     *    who: A string containing the name of the player that is doing the play.
     *    element: The name of the element the player chooses to go with on this Round.
     *
     * Returns:
     *    boolean: true if the play is done, false if it's not because either the player or the
     *        element are not found.
     */
    public boolean play(String who, String element)
    {
        NamedObject player = this.player(who);
        if (player != null)
        {
            Element played = this.element(element);
            if (played != null)
            {
                this.plays.put(player.name(), played);
                return true;
            }
        }
        return false;
    }

    /** Getter for the play a specific player made.
     *   
     * Args:
     *    player: Name of the player to search the play for.
     *
     * Returns:
     *    Element: The Element the player played in this Round, or null if it hasn't played yet.
     */
    public Element playFrom(String player)
    {
        if (player != null)
        {
            return this.plays.get(player);
        }
        return null;
    }

    /** It returns a reference to the player that won the current Round.
     *   
     * Returns:
     *    NamedObject: A reference to the player that won the current Round. If any of the players
     *       hasn't played yet or if the Round is tied, it'll return null.
     */
    public NamedObject winner()
    {
        NamedObject winner = null;
        if (this.plays.size() == 2)
        {
            Element playerAPlay = this.playFrom(this.playerA.name());
            Element playerBPlay = this.playFrom(this.playerB.name());
            if (playerAPlay.isStrongAgainst(playerBPlay))
            {
                winner = this.playerA;
            }
            else if (playerBPlay.isStrongAgainst(playerAPlay))
            {
                winner = this.playerB;
            }
        }
        return winner;
    }

    /** It returns a reference to the player that lost the current Round.
     *   
     * Returns:
     *    NamedObject: A reference to the player that lost the current Round. If any of the players
     *        hasn't played yet or if the Round is tied, it'll return null.
     */
    public NamedObject loser()
    {
        NamedObject winner = this.winner();
        if (this.playerA != winner)
        {
            return this.playerA;
        }
        else if (this.playerB != winner)
        {
            return this.playerB;
        }
        return null;
    }

    private NamedObject playerA = null;
    private NamedObject playerB = null;
    private HashMap<String, Element> elements = null;
    private HashMap<String, Element> plays = null;
}

File: rps_python/round.py
from named_object import NamedObject
from element import Element
from typing import List, Optional


class Round():
    """A class that wraps a round of Rock, Paper, Scissors (or whatever the elements in play are) between two players."""
    def __init__(self, player_a: NamedObject, player_b: NamedObject, available_elements: List[Element]):
        """Initialize the round with the given players and playable elements.
        
        Args:
            player_a: A player in the form of a NamedObject (players do not have any other property).
            player_b: Another player that will play against the first one. Cannot be the same as player_a.
            available_elements: A list of Elements that are available for the players to pick while playing.
        """
        self._player_a = None
        self._player_b = None
        if player_a is not None and player_b is not None and player_a != player_b and player_a.name() != player_b.name():
            self._player_a = player_a
            self._player_b = player_b
        self._plays = {}
        self._elements = {}
        if available_elements is not None:
            for element in available_elements:
                if element is not None and element.name() not in self._elements.keys() and element not in self._elements.values():
                    self._elements[element.name()] = element

    def player(self, by_name: str) -> Optional[NamedObject]:
        """Getter for the players that play against each other on the current Round.
        
        Args:
            by_name: Name of the player to get its associated object.
        
        Returns:
            Optional[NamedObject]: A reference to the player being queried or None if not found.
        """
        if self._player_a.name() == by_name:
            return self._player_a
        elif self._player_b.name() == by_name:
            return self._player_b
        return None

    def players(self) -> List[NamedObject]:
        """The entire list of players that are part of the current Round.
        
        Returns:
            List[NamedObject]: A list of references to players playing on this Round. If a player is missing,
                None will be present in the list.
        """
        return [self._player_a, self._player_b]

    def elements(self) -> List[Element]:
        """Retrieve all Elements that are available for the players to play during this Round.
        
        Returns:
            List[Element]: The list of Elements the players can choose from to play on this Round.
        """
        return list(self._elements.values())

    def element(self, by_name: str) -> Optional[Element]:
        """Analog to player(by_name), allows retrieving a specific Element based on its name.
        
        Args:
            by_name: The name of the Element that should be looked up.

        Returns:
            Optional[Element]: The Element reference, if found, or None if not.
        """
        return self._elements.get(by_name, None)

    def play(self, who: str, element: str) -> bool:
        """Allows a player to perform a play.
        
        Args:
            who: A string containing the name of the player that is doing the play.
            element: The name of the element the player chooses to go with on this Round.

        Returns:
            bool: True if the play is done, False if it's not because either the player or the
                element are not found.
        """
        player = self.player(who)
        if player is not None:
            played = self.element(element)
            if played is not None:
                self._plays[player.name()] = played
                return True
        return False

    def play_from(self, player: str) -> Optional[Element]:
        """Getter for the play a specific player made.
        
        Args:
            player: Name of the player to search the play for.

        Returns:
            Optional[Element]: The Element the player played in this Round, or None if it hasn't played yet.
        """
        if player is not None:
            return self._plays.get(player, None)
        return None

    def winner(self) -> Optional[NamedObject]:
        """It returns a reference to the player that won the current Round.
        
        Returns:
            Optional[NamedObject]: A reference to the player that won the current Round. If any of the players
                hasn't played yet or if the Round is tied, it'll return None.
        """
        winner = None
        if len(self._plays) == 2:
            player_a_play = self.play_from(self._player_a.name())
            player_b_play = self.play_from(self._player_b.name())
            if player_a_play.is_strong_against(player_b_play):
                winner = self._player_a
            elif player_b_play.is_strong_against(player_a_play):
                winner = self._player_b
        return winner

    def loser(self) -> Optional[NamedObject]:
        """It returns a reference to the player that lost the current Round.
        
        Returns:
            Optional[NamedObject]: A reference to the player that lost the current Round. If any of the players
                hasn't played yet or if the Round is tied, it'll return None.
        """
        winner = self.winner()
        if self._player_a != winner:
            return self._player_a
        elif self._player_b != winner:
            return self._player_b
        return None
